File: Array/59. spiral-matrix-ii/59.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> res(n, vector<int>(n, 0));
        int startx = 0, starty = 0;
        int loop = n/2;
        int mid = n/2;
        int offset = 1;
        int count = 1;
        int i, j;
        while(loop--){
            i = startx;
            j = starty;
            for(j; j < n - offset; j++){
                res[i][j] = count;
                count++;
            }
            for(i; i < n - offset; i++){
                res[i][j] = count;
                count++;
            }
            for(j; j>starty; j--){
                res[i][j] = count;
                count++;
            }
            for(i; i>startx; i--){
                res[i][j] = count;
                count++;
            }

            startx++;
            starty++;
            offset++;
        }

        if(n % 2 == 1){
            res[mid][mid] = count;
        }

        return res;

    }
};

File: Array/59. spiral-matrix-ii/59.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        nums = [[0]*n for _ in range(n)]
        startx, starty = 0, 0
        loop, mid = n//2, n//2
        count = 1

        for offset in range(1, loop + 1):
            for i in range(starty, n - offset) :    # 从左至右，左闭右开
                nums[startx][i] = count
                count += 1
            for i in range(startx, n - offset) :    # 从上至下
                nums[i][n - offset] = count
                count += 1
            for i in range(n - offset, starty, -1) : # 从右至左
                nums[n - offset][i] = count
                count += 1
            for i in range(n - offset, startx, -1) : # 从下至上
                nums[i][starty] = count
                count += 1                
            startx += 1         # 更新起始点
            starty += 1

        if n % 2 == 1:
            nums[mid][mid] = count

        return nums
